test/end2end_test.go

```diff
@@ -7128,3 +7128,44 @@ func (s) TestGzipBadChecksum(t *testing.T) {
 		t.Errorf("ss.client.UnaryCall(_) = _, %v\n\twant: _, status(codes.Internal, contains %q)", err, gzip.ErrChecksum)
 	}
 }
+
+// When an RPC is canceled, it's possible that the last Recv() returns before
+// all call options' after are executed.
+func (s) TestCanceledRPCCallOptionRace(t *testing.T) {
+	ss := &stubServer{
+		fullDuplexCall: func(stream testpb.TestService_FullDuplexCallServer) error {
+			err := stream.Send(&testpb.StreamingOutputCallResponse{})
+			if err != nil {
+				return err
+			}
+			<-stream.Context().Done()
+			return nil
+		},
+	}
+	if err := ss.Start(nil); err != nil {
+		t.Fatalf("Error starting endpoint server: %v", err)
+	}
+	defer ss.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	var p peer.Peer
+	stream, err := ss.client.FullDuplexCall(ctx, grpc.Peer(&p))
+	if err != nil {
+		t.Fatalf("_.FullDuplexCall(_) = _, %v", err)
+	}
+	if err := stream.Send(&testpb.StreamingOutputCallRequest{}); err != nil {
+		t.Fatalf("_ has error %v while sending", err)
+	}
+	if _, err := stream.Recv(); err != nil {
+		t.Fatalf("%v.Recv() = %v", stream, err)
+	}
+	cancel()
+	if _, err := stream.Recv(); status.Code(err) != codes.Canceled {
+		t.Fatalf("%v compleled with error %v, want %s", stream, err, codes.Canceled)
+	}
+	// If recv returns before call options are executed, peer.Addr is not set,
+	// fail the test.
+	if p.Addr == nil {
+		t.Fatalf("peer.Addr is nil, want non-nil")
+	}
```

Comment:
Do we need to do this in a loop to have a better chance of stimulating the race?

Reply:
Done